cluster-autoscaler/cloudprovider/clusterapi/clusterapi_machinedeployment.go

```diff
@@ -0,0 +1,136 @@
+	machineDeployment *MachineDeployment
+	maxSize           int
+	minSize           int
+}
+
+var _ scalableResource = (*machineDeploymentScalableResource)(nil)
+
+func (r machineDeploymentScalableResource) ID() string {
+	return path.Join(r.Namespace(), r.Name())
+}
+
+func (r machineDeploymentScalableResource) MaxSize() int {
+	return r.maxSize
+}
+
+func (r machineDeploymentScalableResource) MinSize() int {
+	return r.minSize
+}
+
+func (r machineDeploymentScalableResource) Name() string {
+	return r.machineDeployment.Name
+}
+
+func (r machineDeploymentScalableResource) Namespace() string {
+	return r.machineDeployment.Namespace
+}
+
+func (r machineDeploymentScalableResource) Nodes() ([]string, error) {
+	var result []string
+
+	if err := r.controller.filterAllMachineSets(func(machineSet *MachineSet) error {
+		if machineSetIsOwnedByMachineDeployment(machineSet, r.machineDeployment) {
+			providerIDs, err := r.controller.machineSetProviderIDs(machineSet)
+			if err != nil {
+				return err
+			}
+			result = append(result, providerIDs...)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (r machineDeploymentScalableResource) Replicas() int32 {
+	return pointer.Int32PtrDerefOr(r.machineDeployment.Spec.Replicas, 0)
+}
+
+func (r machineDeploymentScalableResource) SetSize(nreplicas int32) error {
+	u, err := r.controller.dynamicclient.Resource(*r.controller.machineDeploymentResource).Namespace(r.machineDeployment.Namespace).Get(r.machineDeployment.Name, metav1.GetOptions{})
+
+	if err != nil {
+		return err
+	}
+
+	if u == nil {
+		return fmt.Errorf("unknown machineDeployment %s", r.machineDeployment.Name)
+	}
+
+	u = u.DeepCopy()
+	if err := unstructured.SetNestedField(u.Object, int64(nreplicas), "spec", "replicas"); err != nil {
+		return fmt.Errorf("failed to set replica value: %v", err)
+	}
+
+	_, updateErr := r.controller.dynamicclient.Resource(*r.controller.machineDeploymentResource).Namespace(u.GetNamespace()).Update(u, metav1.UpdateOptions{})
+	return updateErr
+}
+
+func (r machineDeploymentScalableResource) MarkMachineForDeletion(machine *Machine) error {
+	u, err := r.controller.dynamicclient.Resource(*r.controller.machineResource).Namespace(machine.Namespace).Get(machine.Name, metav1.GetOptions{})
+
+	if err != nil {
+		return err
+	}
+	if u == nil {
+		return fmt.Errorf("unknown machine %s", machine.Name)
+	}
+
+	u = u.DeepCopy()
+
+	annotations := u.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	annotations[machineDeleteAnnotationKey] = time.Now().String()
+	u.SetAnnotations(annotations)
+
+	_, updateErr := r.controller.dynamicclient.Resource(*r.controller.machineResource).Namespace(u.GetNamespace()).Update(u, metav1.UpdateOptions{})
+	return updateErr
+}
+
+func newMachineDeploymentScalableResource(controller *machineController, machineDeployment *MachineDeployment) (*machineDeploymentScalableResource, error) {
```

Comment:
The intermediate representation of machineDeployment as a custom type of MachineDeployment seems like it could potentially be avoided...  The manipulation I'm seeing of the types appear to be using unstructured, and the majority of the fields being accessed (with the exception of Replicas) appears to have corresponding accessor methods via unstructured.

Reply:
> The intermediate representation of machineDeployment as a custom type of MachineDeployment seems like it could potentially be avoided.

Yes, I agree. 

The switch to unstructured was the cheapest / least disturbing change I could make so that the unit tests would pass again. From a CL perspective it was (in openshift) also easy to review as I only needed to massage places that were previously using a concrete type to now be wrapped. I also think the any new types should be simplified - they need not have TypeMeta, or ObjectMeta, et al.

If were to merge this in its current state a lot of things can be cleaned up. I think it depends on how fast we want to go initially.

Reply:
> The manipulation I'm seeing of the types appear to be using unstructured, and the majority of the fields being accessed (with the exception of Replicas)

(I'm not 100% sure of this) With `Replicas` it may actually be better to look at using the scale subresource.  So, even though access is via unstructured, type safety can still be improved.

I think we still need to poke for `NodeRef` though.

Reply:
Gotcha, makes sense. I think avoiding the use of an intermediate type would help avoid confusion for anyone that is more familiar with the Cluster API codebase rather than the autoscaler codebase. I do agree that it can be cleaned up later, though :)

Reply:
+1 on the idea of using the scale subresource for accessing replicas vs unstructured access.